src/bootstrap/builder.rs

```diff
@@ -435,8 +483,24 @@ impl<'a> ShouldRun<'a> {
         self
     }
 
-    fn pathset_for_path(&self, path: &Path, kind: Kind) -> Option<&PathSet> {
-        self.paths.iter().find(|pathset| pathset.has(path, Some(kind)))
+    /// Given a set of requested paths, return the subset which match the Step for this `ShouldRun`,
+    /// removing the matches from `paths`.
+    ///
+    /// NOTE: this returns multiple PathSets to allow for the possibility of multiple units of work
+    /// within the same step. For example, `test::Crate` allows testing multiple crates in the same
+    /// cargo invocation, which are put into separate sets because they aren't aliases.
+    ///
+    /// The reason we return PathSet instead of PathBuf is to allow for aliases that mean the same thing
+    /// (for now, just `all_krates` and `paths`, but we may want to add an `aliases` function in the future?)
+    fn pathset_for_paths(&self, paths: &mut Vec<&Path>, kind: Kind) -> Vec<PathSet> {
```

Comment:
Similarly to before I'd prefer to avoid an innocuous name like this removing elements from paths. Not sure I have a good suggestion naming wise though. (If we can't come up with anything, let's keep this name for lack of better options).

Reply:
I went with `pathset_for_paths_removing_matches` which seems kinda long but it's only used in one place so I think it's ok to err on the side of being clear.